submissions/AnnaStoyano/frogger/app.js

```diff
@@ -0,0 +1,146 @@
+    }
+    this.increaseLevel();
+}
+Enemy.prototype.checkCollision = function () {
+    let positionX = Math.floor(this.position.x);
+    if (positionX - 50 < this.player.position.x && this.player.position.x < positionX + 50) {
+        if (this.position.y - 55 < this.player.position.y && this.player.position.y < this.position.y + 20) {
+            alert(`You lose!\nYour level is ${this.player.level}`);
+            this.player.restartPosition();
+            this.player.level = 1;
+            return true;
+        }
+    }
+    return false;
+}
+
+Enemy.prototype.increaseLevel = function () {
+    if (this.player.level != this.level) {
+        this.level++;
+    }
+}
+
+
+const Player = function (x, y, speed, sprite) {
+    Entity.call(this, x, y, speed, sprite);
+    this.startX = this.position.x;
+    this.startY = this.position.y;
+}
+
+Player.prototype = Object.create(Entity.prototype);
+Player.prototype.update = function () {
+    this.checkWin();
+    this.checkWallCollision();
+}
+Player.prototype.checkWin = function () {
+    if (this.position.y < -10) {
+        alert(`You win!\nYour level is ${this.level}`);
+        this.level++;
+        this.restartPosition();
+    }
+}
+Player.prototype.restartPosition = function () {
+    this.position.x = this.startX;
+    this.position.y = this.startY;
+}
+
+Player.prototype.checkWallCollision = function () {
+    if (this.position.x < -30) {
+        this.position.x = -20;
+    } else if (this.position.x > canvas.width - 80) {
+        this.position.x = canvas.width - 85;
+    } else if (this.position.y > canvas.height - 5) {
+        this.position.y = canvas.height - 5;
+    }
+}
+
+Player.prototype.handleInput = function (key) {
+    switch (key) {
+        case 'left':
+            this.position.x -= 20;
+            break;
+        case 'right':
+            this.position.x += 20;
+            break;
+        case 'up':
+            this.position.y -= 20;
+            break;
+        case 'down':
+            this.position.y += 20;
+            break;
+    }
+}
+
+const START_POSITION_PLAYER = {
+    X: 200,
+    Y: 404
+};
+const START_POSITION_ENEMY_1 = {
+    X: 0,
+    Y: 120
+};
+const START_POSITION_ENEMY_2 = {
+    X: 15,
+    Y: 220
+};
+const START_POSITION_ENEMY_3 = {
+    X: 30,
+    Y: 50
+};
+const PLAYER_SPEED = 7;
+const ENEMY_1_SPEED = 20;
+const ENEMY_2_SPEED = 60;
+const ENEMY_3_SPEED = 30;
+
+
+const player = new Player(START_POSITION_PLAYER.X, START_POSITION_PLAYER.Y, PLAYER_SPEED, 'images/char-boy.png');
+const enemy1 = new Enemy(START_POSITION_ENEMY_1.X, START_POSITION_ENEMY_1.Y, ENEMY_1_SPEED, 'images/enemy-bug.png', player);
+const enemy2 = new Enemy(START_POSITION_ENEMY_2.X, START_POSITION_ENEMY_2.Y, ENEMY_2_SPEED, 'images/enemy-bug.png', player);
+const enemy3 = new Enemy(START_POSITION_ENEMY_3.X, START_POSITION_ENEMY_3.Y, ENEMY_3_SPEED, 'images/enemy-bug.png', player);
+const allEnemies = [enemy1, enemy2, enemy3];
```

Comment:
Whenever you feel adding an index in a variable name (like `ENEMY_1_SPEED`, `START_POSITION_ENEMY_3` or `enemy2`) this is clear sign that array would be more appropriate.
Imagine you'd be required to add values for dozens of characters.
There is also massive code repetition here (again, think of dozens of characters).
Can you devise an approach that would employ data structure (array of objects containing individual enemy settings)?
You might then efficiently use e.g. `Array.map`.

Hint:
`allEnemies = enemySettings.map(settings => new Enemy(settings.startPosition.X, settings.startPosition.Y, settings.speed, ENEMY_IMAGE, player))`